Odata-docs/webapi-8/tutorials/custom-skiptokenhandler.md

```diff
@@ -0,0 +1,549 @@
+- `context`: Allows you to access information about the current serialization pipeline, such as the request instance, the EDM model, etc.
+
+The method first calls the `GenerateNextPageLink` method of the parent class (`DefaultSkipTokenHandler`) to generate the next link. The method returns `null` if we're on the last page.
+
+```csharp
+var uri = base.GenerateNextPageLink(baseUri, pageSize, instance, context);
+
+if (uri == null)
+{
+    return null;
+}
+```
+
+The next block of code extracts the value of the `$skiptoken` query option from the generated URI. We use the `HttpUtility` class from the `System.Web` namespace to parse the query options.
+
+```csharp
+var queryOptions = HttpUtility.ParseQueryString(uri.Query);
+string skipToken = queryOptions.Get("$skiptoken");
+if (skipToken == null)
+{
+    return uri;
+}
+```
+
+The next block of generates a base64 encoded version of the skip token to replace the original value.
+
+```csharp
+string base64SkipToken = Convert.ToBase64String(Encoding.UTF8.GetBytes(skipToken));
+queryOptions.Set("$skiptoken", base64SkipToken);
+```
+
+Finally, we create a new URI by replacing the original query options with the updated version:
+
+```csharp
+string encodedQuery = queryOptions.ToString();
+UriBuilder builder = new UriBuilder(uri);
+builder.Query = encodedQuery;
+return builder.Uri;
+```
+
+### Testing the new next link
+
+Let's make the following request to see the new skip tokens in action:
+
+```http
+GET /odata/Customers
+```
+
+**Response**:
+
+```json
+**Response**:
+
+```json
+{
+    "@odata.context": "http://localhost:5000/odata/$metadata#Customers",
+    "value": [
+        {
+            "Id": 1,
+            "Name": "Customer 1"
+        },
+        {
+            "Id": 2,
+            "Name": "Customer 2"
+        }
+    ],
+    "@odata.nextLink": "http://localhost:5000/odata/Customers?$skiptoken=SWQtMg%3d%3d"
+}
+```
+
+The `$skiptoken` is now `SWQtMg%3d%3d` instead of `Id-2`.
+
+> [!NOTE]
+> The actual base64 encoding of `Id-2` is `SWQtMg==`, but the final form eventually gets URL encoded to `SWQtMg%3d%3d`.
+
+Let's follow the next link to fetch the next page:
+
+```http
+GET http://localhost:5000/odata/Customers?$skiptoken=SWQtMg%3d%3d
+```
+
+This returns an error response like the following:
+
+```json
+{
+    "error": {
+        "code": "",
+        "message": "The query specified in the URI is not valid. Unable to parse the skiptoken value 'SWQtMg=='. Skiptoken value should always be server generated.",
+    }
+}
+```
+
+The reason we get the error is because our custom handler was not able to process the skip token. The skip token is process in the `ApplyTo` method. At the moment, we are still using the `ApplyTo` method in the `DefaultSkipTokenHandler` class because we have not defined our own. The base `ApplyTo` expects the skip token to be in a specific format, it cannot parse the base64 encoded value.
```

Comment:
```suggestion
The reason we get the error is because our custom handler was not able to process the skip token. The skip token is processed in the `ApplyTo` method. At the moment, we are still using the `ApplyTo` method in the `DefaultSkipTokenHandler` class because we have not defined our own. The base `ApplyTo` expects the skip token to be in a specific format, it cannot parse the base64 encoded value.
```